File: src/ReportGenerator.java
import java.io.*;
import java.math.BigDecimal;
import java.util.*;

/**
 * Created by Sun on 1/5/2020.
 * Consumes CSVs and outputs a CSV for each person
 */
public class ReportGenerator {

    /**
     * maps person name to a list of their donations
     */
    private final Map<String, List<Donation>> donationsByName = new HashMap<>();
    /**
     * what year we are reporting on
     */
    private final int year;

    public ReportGenerator(int year) {
        this.year = year;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        System.out.print("Please enter the desired year: ");
        ReportGenerator reportGenerator = new ReportGenerator(scanner.nextInt());
        reportGenerator.addDonationsFromCSV("donations - check GFCC.csv", Donation.Type.CHECK);
        reportGenerator.addDonationsFromCSV("donations - cash GFCC.csv", Donation.Type.CASH);
        reportGenerator.outputReport();
    }

    private void outputReport() {
        BufferedWriter writer = null;
        try {
            writer = new BufferedWriter(new FileWriter("report_" + year + ".csv"));
            List<String> names = new ArrayList<>(donationsByName.keySet());
            Collections.sort(names);
            for (String name : names) {
                String report = getReportFor(name);
                writer.write(report);
                writer.write(",,,\n");
            }
        } catch (Exception e) {
            e.printStackTrace();
        } finally {
            if (writer != null) {
                try {
                    writer.close();
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
        }
    }

    private String getReportFor(String name) {
        StringBuilder sb = new StringBuilder();
        List<Donation> donations = donationsByName.get(name);
        Collections.sort(donations);
        BigDecimal total = BigDecimal.ZERO;
        for (Donation donation : donations) {
            total = total.add(donation.getAmount());
            sb.append(donation);
            sb.append('\n');
        }
        sb.append(String.format("Total,,\"%s\"\n", Donation.CURRENCY_FORMAT.format(total)));
        return sb.toString();
    }

    private void addDonationsFromCSV(String fileName, Donation.Type type) {
        BufferedReader bufferedReader = null;
        try {
            InputStream inputStream = getClass().getResourceAsStream("resources/" + fileName);
            bufferedReader = new BufferedReader(new InputStreamReader(inputStream));
            String firstLine = bufferedReader.readLine();
            if (!firstLine.equals("Date,Name,Credit")) {
                throw new RuntimeException("first line of CSV '" + fileName + "' should start with column names: Date,Name,Credit");
            }
            while (true) {
                String line = bufferedReader.readLine();
                if (line == null) {
                    break;
                }
                Donation donation = new Donation(line, type);
                if (donation.year != this.year) {
                    continue;
                }
                List<Donation> donations = donationsByName.get(donation.name);
                if (donations == null) {
                    donations = new ArrayList<>();
                }
                donations.add(donation);
                donationsByName.put(donation.name, donations);
            }
        } catch (Exception e) {
            e.printStackTrace();
        } finally {
            if (bufferedReader != null) {
                try {
                    bufferedReader.close();
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
        }
    }

}

File: src/main/kotlin/ReportGenerator.kt
import java.io.BufferedReader
import java.io.File
import java.io.IOException
import java.io.InputStreamReader
import java.util.*

object ReportGenerator {

    /**
     * maps person name to a list of their donations
     */
    private val donationsByName: MutableMap<String, MutableList<Donation>> = HashMap()

    private var year: Int = 0

    @JvmStatic
    fun main(args: Array<String>) {
        val scanner = Scanner(System.`in`)
        print("Please enter the desired year: ")
        year = scanner.nextLine().toInt()
        print("Would you like individual reports or a batch report? (I/B): ")
        val reportType = scanner.nextLine().uppercase()
        print("Would you like to filter people by donation count? (ALL/>#/<#/=#): ")
        val filter = scanner.nextLine().uppercase()
        addDonationsFromCSV("donations - check GFCC.csv", Donation.Type.CHECK)
        addDonationsFromCSV("donations - cash GFCC.csv", Donation.Type.CASH)
        val duplicates = arrayListOf<String>()
        donationsByName.forEach { (name, list) ->
            val set = list.map { "${it.date} ${it.amount}" }.toSet()
            if (list.size != set.size) {
                duplicates.add("duplicates detected for $name")
            }
        }
        val donationsByName = when {
            filter == "ALL" -> donationsByName
            filter.startsWith(">") -> {
                val num = filter.substring(1).toInt()
                donationsByName.filterValues { it.size > num }
            }

            filter.startsWith("<") -> {
                val num = filter.substring(1).toInt()
                donationsByName.filterValues { it.size < num }
            }

            filter.startsWith("=") -> {
                val num = filter.substring(1).toInt()
                donationsByName.filterValues { it.size == num }
            }

            else -> throw IllegalArgumentException("Invalid filter type: $filter")
        }
        print("Would you like DOCX or PDF format? (D/P): ")
        val docFactory: DocumentFactoryProvider = when (val format = scanner.nextLine().uppercase()) {
            "D" -> object : DocumentFactoryProvider {
                override fun provideDocumentFactory(
                    name: String,
                    fein: String,
                    year: Int,
                    filename: String
                ): DocumentFactory = WordDocFactory(name, fein, year, filename)
            }

            "P" -> object : DocumentFactoryProvider {
                override fun provideDocumentFactory(
                    name: String,
                    fein: String,
                    year: Int,
                    filename: String
                ): DocumentFactory = PdfFactory(name, fein, year, filename)
            }

            else -> throw IllegalArgumentException("Invalid format: $format")
        }

        when (reportType) {
            "I" -> donationsByName.forEach { (name, donations) ->
                docFactory.provideDocumentFactory(year = year, filename = "$name (GFCC $year)").run {
                    addFooter()
                    addReport(donations.sorted())
                    writeToFile()
                }
            }

            "B" -> docFactory.provideDocumentFactory(year = year).run {
                addFooter()
                donationsByName.keys.toList().sorted().forEachIndexed { index, name ->
                    addReport(donationsByName[name]!!.sorted())
                    if (index < donationsByName.size - 1) {
                        addPageBreak()
                    }
                }
                writeToFile()
            }

            else -> {
                println("Invalid response. Please try again.")
                return
            }
        }
        println(duplicates.joinToString("\n"))
//        print("Would you like PDFs as well? (Y/N): ")
//        when (scanner.nextLine().uppercase()) {
//            "Y" -> PdfConvertor().convert(docxPaths)
//        }
    }

    private fun addDonationsFromCSV(fileName: String, type: Donation.Type) {
        var bufferedReader: BufferedReader? = null
        try {
            val inputStream = File(System.getProperty("user.home"))
                .resolve("Desktop")
                .resolve(fileName)
                .inputStream()
            bufferedReader = BufferedReader(InputStreamReader(inputStream))
            val firstLine = bufferedReader.readLine()
            if (firstLine != "Date,Name,Credit") {
                throw RuntimeException("first line of CSV '$fileName' should start with column names: Date,Name,Credit")
            }
            while (true) {
                val line = bufferedReader.readLine() ?: break
                val donation = Donation(line, type)
                if (donation.year != this.year) {
                    continue
                }
                val donations = donationsByName[donation.name] ?: ArrayList()
                donations.add(donation)
                donationsByName[donation.name] = donations
            }
        } catch (e: Exception) {
            e.printStackTrace()
        } finally {
            if (bufferedReader != null) {
                try {
                    bufferedReader.close()
                } catch (e: IOException) {
                    e.printStackTrace()
                }
            }
        }
    }
}